### PR TITLE
Perf improvements

### PR DESCRIPTION
**Pull request by gmokki**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,9 @@
   - Fix honoring of `includeCurrentStateVariables` flag in `WorkflowInstanceService.listWorkflowInstances`. This caused major slowness when using bulk workflows.
     To preserve the existing (incorrect) default behaviour in backwards compatible way the default value in `QueryWorkflowInstances.Builder` is changed to `true`. The REST API is unaffected.
     Especially in workflows with many children that use the `StateExecution.getAllChildWorkflows` method the performance impact can be high. Before 7.0.0 release, it is recommended to use `StateExecution.queryChildWorkflows(new QueryWorkflowInstances.Builder().setIncludeCurrentStateVariables(false).build())` if state variables are not needed.
+  - Improve child workflow final state execution speed by caching the parent workflow type
+  - Improve workflow execution type by not scanning for child workflows. If workflow wants to access child workflows they should fetch them using the executor getChildWorklfow methods. Not backwards compatible.
```
Comment (efonsell) [inline CHANGELOG.md]:
```suggestion
  - Improve state execution by not scanning for child workflows instances by default. If access to child workflow instances is needed, they can be fetched using `StateExecution.getChildWorklfow` methods. Not backwards compatible.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,9 @@
...
+  - Fix bulk forkflow concurrency limit calculation.
```
Comment (efonsell) [inline CHANGELOG.md]:
```suggestion
  - Fix bulk workflow concurrency limit calculation.
```

File: nflow-engine/src/test/java/io/nflow/engine/internal/executor/WorkflowStateProcessorTest.java [hunk 3]
```diff
@@ -205,7 +202,7 @@ public void runWorkflowThroughOneSuccessfulState() {
     when(workflowInstancePreProcessor.process(newChildWorkflow)).thenReturn(newChildWorkflow);
     when(workflowInstancePreProcessor.process(newWorkflow)).thenReturn(newWorkflow);
 
-    executor.run();
+    assertTimeoutPreemptively(ofSeconds(5),() -> executor.run());
```
Comment (efonsell) [inline nflow-engine/src/test/java/io/nflow/engine/internal/executor/WorkflowStateProcessorTest.java]: extract method
```suggestion
    assertTimeoutPreemptively(ofSeconds(5), () -> executor.run());
```

File: CHANGELOG.md (hunk 2 above)
Comment (efonsell) [reply CHANGELOG.md]: Was there anything to fix here actually? Could not find that change anymore.

Comment (gmokki) [reply CHANGELOG.md]: My changes were harmful -> removed :)

File: nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 4]
```diff
@@ -133,6 +138,7 @@ public WorkflowInstanceDao(SQLVariants sqlVariants, @NFlow JdbcTemplate nflowJdb
     instanceStateTextLength = env.getProperty("nflow.workflow.instance.state.text.length", Integer.class, -1);
     actionStateTextLength = env.getProperty("nflow.workflow.action.state.text.length", Integer.class, -1);
     stateVariableValueMaxLength = env.getProperty("nflow.workflow.state.variable.value.length", Integer.class, -1);
+    maxWorkflowTypeCacheSize = env.getProperty("nflow.db.type.cache.size", Integer.class, 10_000);
```
Comment (efonsell) [inline nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Rename field and conf property?
```
    workflowInstanceTypeCacheSize = env.getProperty("nflow.workflow.instance.type.cache.size", Integer.class, 10_000);
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -13,6 +13,8 @@
   - Fix honoring of `includeCurrentStateVariables` flag in `WorkflowInstanceService.listWorkflowInstances`. This caused major slowness when using bulk workflows.
     To preserve the existing (incorrect) default behaviour in backwards compatible way the default value in `QueryWorkflowInstances.Builder` is changed to `true`. The REST API is unaffected.
     Especially in workflows with many children that use the `StateExecution.getAllChildWorkflows` method the performance impact can be high. Before 7.0.0 release, it is recommended to use `StateExecution.queryChildWorkflows(new QueryWorkflowInstances.Builder().setIncludeCurrentStateVariables(false).build())` if state variables are not needed.
+  - Improve child workflow final state execution speed by caching the parent workflow type
+  - Optional performance improvement to state execution by not scanning for child workflows instances. If access to child workflow instances is needed, they can be fetched using `StateExecution.getChildWorklfow` methods. To enable set the property `nflow.executor.fetchChildWorkflowIds` to `false`, which is recommended for all. The default functionality will change in next major release.
```
Comment (efonsell) [inline CHANGELOG.md]: Mention new config options in changelog, should they be added to nflow-engine.properties with default values and make the properties required as well?

File: nflow-engine/src/main/resources/nflow-engine.properties [hunk 6]
```diff
@@ -65,5 +65,6 @@ nflow.db.max_pool_size=4
 nflow.db.idle_timeout_seconds=600
 nflow.db.create_on_startup=true
 nflow.db.disable_batch_updates=false
+nflow.db.type.cache.size=10000
```
Comment (efonsell) [inline nflow-engine/src/main/resources/nflow-engine.properties]: Add nflow.executor.fetchChildWorkflowIds with default value here as well, and make it required in code

File: CHANGELOG.md [hunk 7]
```diff
@@ -13,6 +13,8 @@
   - Fix honoring of `includeCurrentStateVariables` flag in `WorkflowInstanceService.listWorkflowInstances`. This caused major slowness when using bulk workflows.
     To preserve the existing (incorrect) default behaviour in backwards compatible way the default value in `QueryWorkflowInstances.Builder` is changed to `true`. The REST API is unaffected.
     Especially in workflows with many children that use the `StateExecution.getAllChildWorkflows` method the performance impact can be high. Before 7.0.0 release, it is recommended to use `StateExecution.queryChildWorkflows(new QueryWorkflowInstances.Builder().setIncludeCurrentStateVariables(false).build())` if state variables are not needed.
+  - Improve child workflow final state execution speed by caching the parent workflow type. Use `nflow.db.type.cache.size` property to tune the size.
```
Comment (efonsell) [inline CHANGELOG.md]: Rename conf property to nflow.workflow.instance.type.cache.size or nflow.workflowInstanceType.cacheSize or something? db type cache size could be anything :)

File: nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java [hunk 8]
```diff
@@ -100,6 +101,8 @@
     stateSaveRetryDelay = env.getRequiredProperty("nflow.executor.stateSaveRetryDelay.seconds", Integer.class);
     stateVariableValueTooLongRetryDelay = env.getRequiredProperty("nflow.executor.stateVariableValueTooLongRetryDelay.minutes",
         Integer.class);
+    // TODO remove flag in 7.x release and default to not fetching child ids (or alternatively, let each step of WorkflowDefinition override what information needs to be fetched)
+    fetchChildWorkflowIds = env.getProperty("nflow.executor.fetchChildWorkflowIds", Boolean.class, true);
```
Comment (efonsell) [inline nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java]: use getRequiredProperty

File: nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 9]
```diff
@@ -867,7 +872,11 @@ public boolean setSignal(long workflowInstanceId, Optional<Integer> signal, Stri
   }
 
   public String getWorkflowInstanceType(long workflowInstanceId) {
-    return jdbc.queryForObject("select type from nflow_workflow where id = ?", String.class, workflowInstanceId);
+    String type = workflowTypeByWorkflowIdCache.computeIfAbsent(workflowInstanceId, id -> jdbc.queryForObject("select type from nflow_workflow where id = ?", String.class, id).intern());
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Can intern() cause memory leaks here?

Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java]: That was a problem in Java 7 and older, but we only support 8+, so not really. Also there will not be that many workflow types and they are not very large strings anyway.

**Pull request merged**